### ISSUE TITLE
Real-time protection causes New-BcContainer to fail

### ISSUE THREAD

**Issue by mtaanquist** [microsoft/navcontainerhelper#2389]
PLEASE DO NOT INCLUDE ANY PASSWORDS OR TOKENS IN YOUR ISSUE!!!

**Describe the issue**
When creating a new BC container with Windows Security's "Real-time protection" enabled, the step "Modifying Service Tier Config File with Instance Specific Settings" causes an error.

At first, I thought this was due to including `-enableTaskScheduler`, but an attempt without also fails.

**Scripts used to create container and cause the issue**
```
New-BcContainer bctest200 `
    -accept_eula `
    -alwaysPull `
    -artifactUrl (Get-BCArtifactUrl -type OnPrem -country DK -version 20) `
    -assignPremiumPlan `
    -auth NavUserPassword `
    -doNotExportObjectsToText `
    -enableTaskScheduler `
    -isolation process `
    -licenseFile "BC 20 DEV.flf" `
    -updateHosts 
```

**Full output of scripts**
```
BcContainerHelper is version 3.0.4
BcContainerHelper is running as administrator
Host is Microsoft Windows 11 Enterprise - 21H2
Docker Client Version is 20.10.13
Docker Server Version is 20.10.13
Fetching all docker images
Fetching all docker volumes
Pulling image mcr.microsoft.com/businesscentral:10.0.20348.587
10.0.20348.587: Pulling from businesscentral
Digest: sha256:c9ebf4f5a6d5d817aa883790f8eaac4a463c1412fc7be73573a81d7858c2e999
Status: Image is up to date for mcr.microsoft.com/businesscentral:10.0.20348.587
mcr.microsoft.com/businesscentral:10.0.20348.587
Using image mcr.microsoft.com/businesscentral:10.0.20348.587
Creating Container bctest200
Style: onprem
Multitenant: No
Version: 20.0.37253.38230
Platform: 20.0.37114.38150
Generic Tag: 1.0.2.1
Container OS Version: 10.0.20348.587 (ltsc2022)
Host OS Version: 10.0.22000.556 (21H2)
Using process isolation
Using locale da-DK
Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
Using license file C:\PathTo\BC 20 DEV.flf
Additional Parameters:
--env customNavSettings=EnableTaskScheduler=True
Files in C:\ProgramData\BcContainerHelper\Extensions\bctest200\my:
- AdditionalOutput.ps1
- license.flf
- MainLoop.ps1
- SetupNavUsers.ps1
- SetupVariables.ps1
- updatehosts.ps1
Creating container bctest200 from image mcr.microsoft.com/businesscentral:10.0.20348.587
effeb7b3a31adccb94e18a1ced439273ab2cd89737a3e81472dd9be522ffbc45
Waiting for container bctest200 to be ready
Using artifactUrl https://bcartifacts.azureedge.net/onprem/20.0.37253.38230/dk
Using installer from C:\Run\150-new
Installing Business Central
Installing from artifacts
Starting Local SQL Server
Starting Internet Information Server
Copying Service Tier Files
c:\dl\onprem\20.0.37253.38230\platform\ServiceTier\Program Files
c:\dl\onprem\20.0.37253.38230\platform\ServiceTier\System64Folder
Copying PowerShell Scripts
c:\dl\onprem\20.0.37253.38230\platform\WindowsPowerShellScripts\Cloud\NAVAdministration
c:\dl\onprem\20.0.37253.38230\platform\WindowsPowerShellScripts\WebSearch
Copying dependencies
Copying ReportBuilder
Importing PowerShell Modules
Determining Database Collation from c:\dl\onprem\20.0.37253.38230\dk\database\Demo Database BC (20-0).bak
Initializing...
Setting host.docker.internal to 192.168.2.33 in container hosts file (copy from host hosts file)
Setting gateway.docker.internal to 192.168.2.33 in container hosts file (copy from host hosts file)
Setting kubernetes.docker.internal to 127.0.0.1 in container hosts file (copy from host hosts file)
Setting host.containerhelper.internal to 172.27.240.1 in container hosts file
Starting Container
Hostname is bctest200
PublicDnsName is bctest200
Using NavUserPassword Authentication
Starting Local SQL Server
Starting Internet Information Server
Creating Self Signed Certificate
Self Signed Certificate Thumbprint 1BCC7C917237F9736D561567D46F2DCBD4EA5D40
DNS identity bctest200
Modifying Service Tier Config File with Instance Specific Settings
Modifying Service Tier Config File with settings from environment variable
Setting EnableTaskScheduler to True
Service Tier doesn't exist / is not installed
at <ScriptBlock>, C:\Run\navstart.ps1: line 153
at <ScriptBlock>, C:\Run\start.ps1: line 363
at <ScriptBlock>, <No file>: line 1Error
Using artifactUrl https://bcartifacts.azureedge.net/onprem/20.0.37253.38230/dk
Using installer from C:\Run\150-new
Installing Business Central
Installing from artifacts
Starting Local SQL Server
Starting Internet Information Server
Copying Service Tier Files
c:\dl\onprem\20.0.37253.38230\platform\ServiceTier\Program Files
c:\dl\onprem\20.0.37253.38230\platform\ServiceTier\System64Folder
Copying PowerShell Scripts
c:\dl\onprem\20.0.37253.38230\platform\WindowsPowerShellScripts\Cloud\NAVAdministration
c:\dl\onprem\20.0.37253.38230\platform\WindowsPowerShellScripts\WebSearch
Copying dependencies
Copying ReportBuilder
Importing PowerShell Modules
Determining Database Collation from c:\dl\onprem\20.0.37253.38230\dk\database\Demo Database BC (20-0).bak
Initializing...
Setting host.docker.internal to 192.168.2.33 in container hosts file (copy from host hosts file)
Setting gateway.docker.internal to 192.168.2.33 in container hosts file (copy from host hosts file)
Setting kubernetes.docker.internal to 127.0.0.1 in container hosts file (copy from host hosts file)
Setting host.containerhelper.internal to 172.27.240.1 in container hosts file
Starting Container
Hostname is bctest200
PublicDnsName is bctest200
Using NavUserPassword Authentication
Starting Local SQL Server
Starting Internet Information Server
Creating Self Signed Certificate
Self Signed Certificate Thumbprint 1BCC7C917237F9736D561567D46F2DCBD4EA5D40
DNS identity bctest200
Modifying Service Tier Config File with Instance Specific Settings
Modifying Service Tier Config File with settings from environment variable
Setting EnableTaskScheduler to True
Service Tier doesn't exist / is not installed
at <ScriptBlock>, C:\Run\navstart.ps1: line 153
at <ScriptBlock>, C:\Run\start.ps1: line 363
at <ScriptBlock>, <No file>: line 1
New-BcContainer Telemetry Correlation Id: 019bf0f9-6932-4e36-b127-385d4e6777d8
Initialization of container bctest200 failed
At C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\3.0.4\ContainerHandling\Wait-NavContainerReady.ps1:42 c
har:17
+ ...             throw "Initialization of container $containerName failed" ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Initialization ...ctest200 failed:String) [], RuntimeException
    + FullyQualifiedErrorId : Initialization of container bctest200 failed
...
```

If I disable real-time protection in Windows Security, it runs through fine:
```
BcContainerHelper is version 3.0.4
BcContainerHelper is running as administrator
Host is Microsoft Windows 11 Enterprise - 21H2
Docker Client Version is 20.10.13
Docker Server Version is 20.10.13
Removing container bctest200
Removing bctest200 from container hosts file
Removing bctest200-* from container hosts file
Removing C:\ProgramData\BcContainerHelper\Extensions\bctest200
Fetching all docker images
Fetching all docker volumes
Pulling image mcr.microsoft.com/businesscentral:10.0.20348.587
10.0.20348.587: Pulling from businesscentral
Digest: sha256:c9ebf4f5a6d5d817aa883790f8eaac4a463c1412fc7be73573a81d7858c2e999
Status: Image is up to date for mcr.microsoft.com/businesscentral:10.0.20348.587
mcr.microsoft.com/businesscentral:10.0.20348.587
Using image mcr.microsoft.com/businesscentral:10.0.20348.587
Creating Container bctest200
Style: onprem
Multitenant: No
Version: 20.0.37253.38230
Platform: 20.0.37114.38150
Generic Tag: 1.0.2.1
Container OS Version: 10.0.20348.587 (ltsc2022)
Host OS Version: 10.0.22000.556 (21H2)
Using process isolation
Using locale da-DK
Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
Using license file C:\PathTo\BC 20 DEV.flf
Files in C:\ProgramData\BcContainerHelper\Extensions\bctest200\my:
- AdditionalOutput.ps1
- license.flf
- MainLoop.ps1
- SetupNavUsers.ps1
- SetupVariables.ps1
- updatehosts.ps1
Creating container bctest200 from image mcr.microsoft.com/businesscentral:10.0.20348.587
7240fced61b0ed7ce6575fb1d5f40ca58b5d570229a0de48756dce1b0cfc44cc
Waiting for container bctest200 to be ready
Using artifactUrl https://bcartifacts.azureedge.net/onprem/20.0.37253.38230/dk
Using installer from C:\Run\150-new
Installing Business Central
Installing from artifacts
Starting Local SQL Server
Starting Internet Information Server
Copying Service Tier Files
c:\dl\onprem\20.0.37253.38230\platform\ServiceTier\Program Files
c:\dl\onprem\20.0.37253.38230\platform\ServiceTier\System64Folder
Copying PowerShell Scripts
c:\dl\onprem\20.0.37253.38230\platform\WindowsPowerShellScripts\Cloud\NAVAdministration
c:\dl\onprem\20.0.37253.38230\platform\WindowsPowerShellScripts\WebSearch
Copying dependencies
Copying ReportBuilder
Importing PowerShell Modules
Determining Database Collation from c:\dl\onprem\20.0.37253.38230\dk\database\Demo Database BC (20-0).bak
Changing Database Server Collation to Danish_Greenlandic_100_CI_AS
Restoring CRONUS Demo Database
Setting CompatibilityLevel for CRONUS on localhost\SQLEXPRESS
Modifying Business Central Service Tier Config File for Docker
Creating Business Central Service Tier
Installing SIP crypto provider: 'C:\Windows\System32\NavSip.dll'
Copying Web Client Files
c:\dl\onprem\20.0.37253.38230\platform\WebClient\Microsoft Dynamics NAV
Copying Client Files
c:\dl\onprem\20.0.37253.38230\platform\LegacyDlls\program files\Microsoft Dynamics NAV
c:\dl\onprem\20.0.37253.38230\platform\LegacyDlls\program files\Microsoft Dynamics NAV
c:\dl\onprem\20.0.37253.38230\platform\LegacyDlls\systemFolder
Copying ModernDev Files
c:\dl\onprem\20.0.37253.38230\platform
c:\dl\onprem\20.0.37253.38230\platform\ModernDev\program files\Microsoft Dynamics NAV
Copying additional files
Copying ConfigurationPackages
C:\dl\onprem\20.0.37253.38230\dk\ConfigurationPackages
Copying Test Assemblies
C:\dl\onprem\20.0.37253.38230\platform\Test Assemblies
Copying Applications
C:\dl\onprem\20.0.37253.38230\dk\Applications
Starting Business Central Service Tier
Importing license file
Stopping Business Central Service Tier
Installation took 118 seconds
Installation complete
Initializing...
Setting host.docker.internal to 192.168.2.33 in container hosts file (copy from host hosts file)
Setting gateway.docker.internal to 192.168.2.33 in container hosts file (copy from host hosts file)
Setting kubernetes.docker.internal to 127.0.0.1 in container hosts file (copy from host hosts file)
Setting host.containerhelper.internal to 172.27.240.1 in container hosts file
Starting Container
Hostname is bctest200
PublicDnsName is bctest200
Using NavUserPassword Authentication
Creating Self Signed Certificate
Self Signed Certificate Thumbprint 9816818F51640A3D996DC501F2B7768B28590784
DNS identity bctest200
Modifying Service Tier Config File with Instance Specific Settings
Starting Service Tier
Registering event sources
Creating DotNetCore Web Server Instance
Using license file 'c:\run\my\license.flf'
Import License
Creating http download site
Setting SA Password and enabling SA
Creating mt as SQL User and add to sysadmin
Creating SUPER user
Assign Premium plan for MT
Container IP Address: 172.27.247.235
Container Hostname  : bctest200
Container Dns Name  : bctest200
Web Client          : http://bctest200/BC/
Dev. Server         : http://bctest200
Dev. ServerInstance : BC
Setting bctest200 to 172.27.247.235 in host hosts file

Files:
http://bctest200:8080/ALLanguage.vsix

Container Total Physical Memory is 31.6Gb
Container Free Physical Memory is 13.6Gb

Initialization took 33 seconds
Ready for connections!
Reading CustomSettings.config from bctest200
Creating Desktop Shortcuts for bctest200
Container bctest200 successfully created

Use:
Get-BcContainerEventLog -containerName bctest200 to retrieve a snapshot of the event log from the container
Get-BcContainerDebugInfo -containerName bctest200 to get debug information about the container
Enter-BcContainer -containerName bctest200 to open a PowerShell prompt inside the container
Remove-BcContainer -containerName bctest200 to remove the container again
docker logs bctest200 to retrieve information about URL's again
```

**Screenshots**
Disabled Real-time protection:
![image](https://user-images.githubusercontent.com/12089874/161937687-09acca04-aa05-4cb4-a908-3da464cd007f.png)

**Additional context**
Versions:
- Windows 11 21H2 (OS Build 22000.556)
- BcContainerHelper 3.0.4
- Docker Desktop 4.6.1 (76265)

**Comment by freddydk** [comment]
I am running the same build with the same settings, but don't have the same problem.
I assume that -isolation hyperv solves the issue.
Are you using controlled folder access?
Did you try to add c:\programdata\bccontainerhelper to exclusions?

Thanks

**Comment by git-no** [comment]
Maybe related to #2374

**Comment by hiddentextbox** [comment]
I ran into the same issue.  had to turn off real time protection. I will try to add the bccontainerhelper to exclusion list.  thanks @freddydk  and @mtaanquist

**Comment by mtaanquist** [comment]
> I am running the same build with the same settings, but don't have the same problem. I assume that -isolation hyperv solves the issue. Are you using controlled folder access? Did you try to add c:\programdata\bccontainerhelper to exclusions?
> 
> Thanks

Apologies for the late reply, I've been on vacation. As mentioned also in #2374, adding the exclusion folder made no difference, and it still fails at the same time. I also tried to add an exclusion for `C:\bcartifacts.cache`, but this didn't make any changes either.

However, browsing through the Event Viewer to see if I could find anything that would be useful, I saw two error entries from docker, with the following messages:

```
exec's CreateProcess() failed [error=container 8e5b896daea74ea64595ac6ba26846dce7aa30e21f0e0a30ce0e7e86b9ef6562 encountered an error during hcsshim::System::CreateProcess: failure in a Windows system call: The user name or password is incorrect. (0x52e) exec=d6c15f89c604492bbe86d84351adcf5032a566b40b5b85db92fa4592c768ba14 container=8e5b896daea74ea64595ac6ba26846dce7aa30e21f0e0a30ce0e7e86b9ef6562 module=libcontainerd namespace=moby]
```

```
Error running exec d6c15f89c604492bbe86d84351adcf5032a566b40b5b85db92fa4592c768ba14 in container: container 8e5b896daea74ea64595ac6ba26846dce7aa30e21f0e0a30ce0e7e86b9ef6562 encountered an error during hcsshim::System::CreateProcess: failure in a Windows system call: The user name or password is incorrect. (0x52e)
```

I have a telemetry ID from it if that's helpful: 
```
New-BcContainer Telemetry Correlation Id: b505fcae-3985-4005-9072-e66c9fc3b7e4
```

**Comment by mtaanquist** [comment]
A bit more information, from more trial and error.

Excluding `docker.exe` and `Docker Desktop.exe` processes from Windows Defender does not make a difference. No difference when adding `C:\ProgramData\Docker\*` to the exclusions either.

When it fails on creation, and I get the event log from the container, the following error is listed:
```xml
<Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
  <System>
    <Provider Name="Microsoft-Windows-CertificateServicesClient-CertEnroll" Guid="{54164045-7C50-4905-963F-E5BC1EEF0CCA}" EventSourceName="CertEnroll" /> 
    <EventID Qualifiers="33370">57</EventID> 
    <Version>0</Version> 
    <Level>2</Level> 
    <Task>0</Task> 
    <Opcode>0</Opcode> 
    <Keywords>0x80000000000000</Keywords> 
    <TimeCreated SystemTime="2022-04-20T11:14:44.7979472Z" /> 
    <EventRecordID>860</EventRecordID> 
    <Correlation /> 
    <Execution ProcessID="0" ThreadID="0" /> 
    <Channel>Application</Channel> 
    <Computer>bctest</Computer> 
    <Security UserID="S-1-5-93-2-1" /> 
  </System>
  <EventData>
    <Data Name="Context">Local system</Data> 
    <Data Name="Provider">Microsoft Passport Key Storage Provider</Data> 
    <Data Name="Failure">NCryptOpenStorageProvider</Data> 
    <Data Name="ErrorCode">Provider DLL could not be found. 0x8009001e (-2146893794 NTE_PROV_DLL_NOT_FOUND)</Data> 
  </EventData>
</Event>
```

I have also updated to latest Docker Desktop version (now 4.7.1) with no change in the inforamtion.

**Comment by epernst** [comment]
Had this issue for several months now. Ever since I upgraded to Windows 10 21h2, then I had to disable real-time protection to create a new container. Once the container is created then I can enable it again.
And I can see I'm not the only one: https://businesscentralgeek.com/my-journey-in-al-go-for-github

**Comment by mtaanquist** [comment]
As mentioned in #2434, freddydk has put up a new image which solves this issue.

Run your New-BcContainer script with the following additional argument, and it should succeed: `-useGenericImage "$(Get-BestGenericImageName)-dev"`

Eventually, I'm sure it'll be rolled into the primary image, but until then, this is a decent work-around.

**Comment by freddydk** [comment]
Generic image 1.0.2.5 has now shipped together with BcContainerHelper 3.0.8
